### PR TITLE
[SMTChecker] Create SMTSolver option in isoltest

### LINKED ISSUE #4735: Deal with potential non-determinism in SMT checker tests.
The results of the SMT checker tests may depend on the enabled solvers and (worse) on the available CPU power.

For the former problem, having test expectations specific to solvers should be sufficient, but this still wouldn't solve the latter problem.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/4735

Depends on https://github.com/ethereum/solidity/pull/7880

### CODE REVIEW

**Comment by leonardoalt** [comment]
@ekpyron It seems like increasing the limit helped z3 solve the nonlinear problems as well.

File: test/libsolidity/SyntaxTest.cpp [hunk 1]
```diff
@@ -207,6 +157,71 @@ void SyntaxTest::printSource(ostream& _stream, string const& _linePrefix, bool _
...
+	compiler().setOptimiserSettings(
+		m_optimiseYul ?
+		OptimiserSettings::full() :
+		OptimiserSettings::minimal()
+	);
+}
+
+void SyntaxTest::parseAndAnalyze()
+{
+	if (compiler().parse())
```
Comment (Marenz) [inline test/libsolidity/SyntaxTest.cpp]: why not `if (compiler().parsen) && compiler.analyze())`

**Review by Marenz — CHANGES_REQUESTED** [review]
LGTM, only a minor comment

File: test/libsolidity/SyntaxTest.cpp (hunk 1 above)
Comment (leonardoalt) [reply test/libsolidity/SyntaxTest.cpp]: I just copied that while splitting the content into new functions

Comment (erak) [reply test/libsolidity/SyntaxTest.cpp]: Combining them seems to make sense.

File: libsolidity/formal/ModelChecker.h [hunk 2]
```diff
@@ -59,6 +59,10 @@ class ModelChecker
 	/// the constructor.
 	std::vector<std::string> unhandledQueries();
 
+	/// @returns which SMT solvers are available
```
Comment (erak) [inline libsolidity/formal/ModelChecker.h]: `///  Returns SMT solvers that are available via the C++ API.`

File: test/libsolidity/SMTCheckerTest.cpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+		if (choice == "any")
+			m_enabledSolvers = smt::SMTSolverChoice::All();
+		else if (choice == "z3")
+			m_enabledSolvers = smt::SMTSolverChoice::Z3();
+		else if (choice == "cvc4")
+			m_enabledSolvers = smt::SMTSolverChoice::CVC4();
+		else if (choice == "none")
+			m_enabledSolvers = smt::SMTSolverChoice::None();
+		else
+			BOOST_THROW_EXCEPTION(runtime_error("Invalid SMT solver choice."));
```
Comment (erak) [inline test/libsolidity/SMTCheckerTest.cpp:43]: Could use uniform initialization: `runtime_error{"Invalid SMT solver choice."}`

File: test/libsolidity/SMTCheckerTest.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		return std::make_unique<SMTCheckerTest>(_config.filename, _config.evmVersion);
+	}
+	SMTCheckerTest(std::string const& _filename, langutil::EVMVersion _evmVersion);
+
+	TestResult run(std::ostream& _stream, std::string const& _linePrefix = "", bool _formatted = false) override;
+
+	bool validateSettings(langutil::EVMVersion _evmVersion) override;
+
+protected:
+	smt::SMTSolverChoice m_enabledSolvers;
```
Comment (erak) [inline test/libsolidity/SMTCheckerTest.h:50]: Perhaps add a comment about the member being set to whatever is specified in the options?

**Review by erak — COMMENTED** [review]
Just left some minor comments. Looks good otherwise!

File: test/libsolidity/SMTCheckerTest.cpp (hunk 3 above)
Comment (leonardoalt) [reply test/libsolidity/SMTCheckerTest.cpp:43]: Instead of BOOST...?

Comment (leonardoalt) [reply test/libsolidity/SMTCheckerTest.cpp:43]: Otherwise, what would be the reason for the change?

**Comment by leonardoalt** [comment]
@erak @Marenz updated

**Review by Marenz — DISMISSED** [review]

**Comment by leonardoalt** [comment]
@Marenz I'll squash the last commit and ask you to approve again

**Comment by leonardoalt** [comment]
@Marenz done

**Review by Marenz — APPROVED** [review]

**Pull request merged**